### PR TITLE
Json writer

### PR DESCRIPTION
**Pull request by gbalduzz**
As I was annoyed while debugging from all the exceptions thrown by the json reader, I wrote my own parser. Improvement

- Instead of throwing when reading a non-existing field, the execute method returns false and the object is unmodified.
- Allows to reopen and overwrite existing groups and entries, allowing to use the same interface of the HDF5 reder/writer, run the same suit of test, and archive true polimorphism in the DcaLoop class.
- The data is stored internally as a tree with strings as leaves. The data is converted to the desired type when execute is called. This is much more readable from a debugger.
- Allows to easily add support for custom data types. 

TODO: have a more readable output for matrices and function if required.

### CODE REVIEW

**Comment by gbalduzz** [comment]
retest this, please

**Comment by gbalduzz** [comment]
retest this, please

**Review by PDoakORNL — APPROVED** [review]
Ok this looks good enough to me.  I'm going to file an issue to keep a couple small changes we probably will want to make in the future.

File: include/dca/io/json/json_reader.hpp [hunk 1]
```diff
@@ -1,329 +1,146 @@
...
+  constexpr static bool is_writer = false;
 
-  template <typename scalartype, typename domain_type>
-  void execute(std::string name, func::function<std::complex<scalartype>, domain_type>& f,
-               const JsonAccessor& current_result, std::size_t index);
+  // Reads the variable with given name from the current group into the object "obj".
+  // If the variable or the currently open group does not exist, or the internal string
+  // representation is not compatible with the desired type "T" returns false and leaves "obj"
+  // unmodified.
+  template <class T>
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/io/json/json_reader.hpp:52]: Agreed that there should be no exception in a regular execution of a code.

**Pull request merged**